snuba/subscriptions/data.py

```diff
@@ -166,10 +180,71 @@ class SnQLSubscriptionData(SubscriptionData):
     def build_request(
         self, dataset: Dataset, timestamp: datetime, offset: Optional[int], timer: Timer
     ) -> Request:
-        # TODO: Parse the query without doing the full validation
-        # TODO: Add time range conditions
-        # TODO: add offset condition
-        raise NotImplementedError
+        schema = RequestSchema.build_with_extensions(
+            {}, SubscriptionRequestSettings, Language.SNQL,
+        )
+
+        def add_conditions(query: Union[CompositeQuery[Entity], Query]) -> None:
+            # TODO: Support composite queries with multiple entities.
+            from_clause = query.get_from_clause()
+            if not isinstance(from_clause, Entity):
+                raise InvalidSubscriptionError("Only simple queries are supported")
+            entity = get_entity(from_clause.key)
+            required_timestamp_column = entity.required_time_column
+            if required_timestamp_column is None:
+                raise InvalidSubscriptionError(
+                    "Entity must have a timestamp column for subscriptions"
+                )
+
+            conditions_to_add: List[Expression] = [
+                binary_condition(
+                    ConditionFunctions.EQ,
+                    Column(None, None, "project_id"),
+                    Literal(None, self.project_id),
+                ),
+                binary_condition(
+                    ConditionFunctions.GTE,
+                    Column(None, None, required_timestamp_column),
+                    Literal(None, (timestamp - self.time_window)),
+                ),
+                binary_condition(
+                    ConditionFunctions.LT,
+                    Column(None, None, required_timestamp_column),
+                    Literal(None, timestamp),
+                ),
+            ]
+
+            if offset is not None:
+                conditions_to_add.append(
+                    binary_condition(
+                        ConditionFunctions.LTE,
+                        FunctionCall(
+                            None,
+                            "ifnull",
+                            (Column(None, None, "offset"), Literal(None, 0)),
+                        ),
+                        Literal(None, offset),
+                    )
+                )
+
+            condition = query.get_condition()
+            if condition:
+                conditions_to_add = [
+                    *get_first_level_and_conditions(condition),
+                    *conditions_to_add,
+                ]
+
+            query.set_ast_condition(combine_and_conditions(conditions_to_add))
```

Comment:
I think it would be more performant to do `query.set_ast_condition(binary_condition("AND", condition, combine_and_conditions(conditions_to_add))` instead of expanding and collapsing all the conditions.